invokeai/app/api/routers/download_queue.py

```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2023 Lincoln D. Stein
+"""FastAPI route for the download queue."""
+
+from typing import List, Optional
+
+from fastapi import Body, Path, Response
+from fastapi.routing import APIRouter
+from pydantic.networks import AnyHttpUrl
+from starlette.exceptions import HTTPException
+
+from invokeai.app.services.download import (
+    DownloadJob,
+    UnknownJobIDException,
+)
+
+from ..dependencies import ApiDependencies
+
+download_queue_router = APIRouter(prefix="/v1/download_queue", tags=["download_queue"])
```

Comment:
I'm currently treating the download service as a separate service that could potentially be used outside the installation system to download IP and controlnet images, community workflows, or whatever we might want to do some day. No actual use caess right now, so we could make this a behind-the-scenes service that doesn't have a route API.

However, there's also (minor) subtly here that we should (briefly) discuss. An install job may trigger multiple download jobs, since a diffusers model usually has multiple individual files to download. When reporting progress events, do we want to report %progress from the standpoint of the install job or the individual download jobs? 

What I've been planning to do is to have separate installer and downloader progress events. The installer progress event reports progress towards the total bytes from all the downloaded files combined, and this is what the user interface shows. The download queue also emits events, but these are for the individual files comprising the model and probably shouldn't be shown in the UI. This would leave open the possibility of a display like this one sometime in the future:

```
v XpUCT/Deliberate =================================>.................................. 80%
   config.json =======================================================================>100%
   pytorch_model.safetensors ===========>...............................................30%
   model.safetensors......................................................................0%
```

Reply:
I like showing the different parts! How about something like this for the event payload:

```py
from pydantic import BaseModel


class ProgressBase(BaseModel):
    label: str
    bytes_downloaded: int
    bytes_total: int


class InstallJobProgress(ProgressBase):
    parts: list[ProgressBase]


job_progress = InstallJobProgress(
    label="XpUCT/Deliberate",
    bytes_downloaded=100,
    bytes_total=1000,
    parts=[
        ProgressBase(
            label="config.json",
            bytes_downloaded=10,
            bytes_total=100,
        ),
        ProgressBase(
            label="pytorch_model.safetensors",
            bytes_downloaded=20,
            bytes_total=200,
        ),
        ProgressBase(
            label="model.safetensors",
            bytes_downloaded=20,
            bytes_total=200,
        ),
    ],
)
```

Reply:
Sure! I will put something like that into the installer progress payload. That would be in the next PR where I integrate the installer with the download queue.